File: solutions/lua/yacht/1/yacht.lua
---@param xs integer[]
---@return integer[]
local function sorted(xs)
	local ys = {}
	for _, v in pairs(xs) do
		table.insert(ys, v)
	end
	table.sort(ys)
	return ys
end

---@generic K, V
---@param xm { [K]: V }
---@param pred fun(k: K, v: V): boolean
---@return boolean
local function all(xm, pred)
	for k, v in pairs(xm) do
		if not pred(k, v) then
			return false
		end
	end
	return true
end

---@generic T
---@param set T[]
---@return { [T]: integer }
local function frequencies(set)
	local res = {}
	for _, x in pairs(set) do
		res[x] = (res[x] or 0) + 1
	end
	return res
end

---@generic T
---@param set T[]
---@param pred fun(v: T): boolean
---@return boolean
local function any(set, pred)
	for _, v in pairs(set) do
		if pred(v) then
			return true
		end
	end
	return false
end


---@generic T
---@param x T
---@param set T[]
---@return boolean
local function memberof(x, set)
	for _, v in pairs(set) do
		if x == v then
			return true
		end
	end
	return false
end

---@param xs integer[]
---@return integer
local function sum(xs)
	local acc = 0
	for _, x in pairs(xs) do
		acc = acc + x
	end
	return acc
end

---@generic T
---@param xs T[]
---@param pred fun(x: T): boolean
---@return integer
local function count(xs, pred)
	local acc = 0
	for _, x in pairs(xs) do
		if pred(x) then
			acc = acc + 1
		end
	end
	return acc
end

local yacht = {}

local cat_to_num = {
	['ones'] = 1,
	['twos'] = 2,
	['threes'] = 3,
	['fours'] = 4,
	['fives'] = 5,
	['sixes'] = 6,
}

local function is_little_straight(dice)
	return all(sorted(dice), function(i, d) return d == i end)
end

local function is_big_straight(dice)
	return all(sorted(dice), function(i, d) return d == (i+1) end)
end

local function is_yacht(dice)
	local prev = dice[1]
	for i, d in ipairs(dice) do
		if i ~= 1 then
			if d ~= prev then
				return false
			end
			prev = d
		end
	end
	return true
end

local function has_same(dice, count)
	local counts = frequencies(dice)
	return any(counts, function(v) return v == count end)
end

local function has_same_or_bigger(dice, count)
	local counts = frequencies(dice)
	return any(counts, function(v) return v >= count end)
end

local function is_category(dice, category)
	if category == 'full house' then
		return has_same(dice, 2) and has_same(dice, 3)
	elseif category == 'four of a kind' then
		return has_same_or_bigger(dice, 4)
	elseif category == 'little straight' then
		return is_little_straight(dice)
	elseif category == 'big straight' then
		return is_big_straight(dice)
	elseif category == 'yacht' then
		return is_yacht(dice)
	else
		return true
	end
end

---@return integer
local function count_same(dice, count)
	for d, v in pairs(frequencies(dice)) do
		if v >= count then
			return d
		end
	end

	return 0
end

---@param dice integer[]
---@return integer
function yacht.score(dice, category)
	if not is_category(dice, category) then
		return 0
	end

	if memberof(category, {'ones', 'twos', 'threes', 'fours', 'fives', 'sixes'}) then
		local num = cat_to_num[category]
		local acc = count(dice, function(d) return d == num end)
		return num * acc
	elseif memberof(category, {'full house', 'choice'}) then
		return sum(dice)
	elseif category == 'four of a kind' then
		return count_same(dice, 4) * 4
	elseif category == 'choice' then
		return sum(dice)
	elseif memberof(category, {'little straight', 'big straight'}) then
		return 30
	elseif category == 'yacht' then
		return 50
	else
		assert(false)
		return -1
	end
end

return yacht

File: solutions/c/yacht/1/yacht.c
#include "yacht.h"

#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>

static int count(dice_t dice, int num) {
	int acc = 0;
	for (size_t i = 0; i < 5; i++)
		if (dice.faces[i] == num)
			acc++;
	return acc;
}

static int sum(dice_t dice) {
	int acc = 0;
	for (size_t i = 0; i < 5; i++)
		acc += dice.faces[i];
	return acc;
}

static int comp(const void *a, const void *b) {
	return *(int*)a - *(int*)b;
}

static bool is_little_straight(dice_t dice) {
	int sorted_faces[5];
	memcpy(sorted_faces, dice.faces, 5 * sizeof(int));
	qsort(sorted_faces, 5, sizeof(int), comp);
	for (size_t i = 0; i < 5; i++)
		if (sorted_faces[i] != (int)(i+1))
			return false;
	return true;
}

static bool is_big_straight(dice_t dice) {
	int sorted_faces[5];
	memcpy(sorted_faces, dice.faces, 5 * sizeof(int));
	qsort(sorted_faces, 5, sizeof(int), comp);
	for (size_t i = 0; i < 5; i++)
		if (sorted_faces[i] != (int)(i+2))
			return false;
	return true;
}

static bool is_yacht(dice_t dice) {
	for (size_t i = 0; i < (5-1); i++)
		if (dice.faces[i] != dice.faces[i+1])
			return false;
	return true;
}

typedef struct {
  int first;
	int second;
} IntPair;

typedef struct {
	IntPair* pairs;
	int count;
} IntMap;

static IntMap frequencies(int set[static 5]) {
	IntPair* freqs = malloc(sizeof(IntPair) * 5);
	int freq_count = 0;
	for (size_t i = 0; i < 5; i++) {
		int found = -1;
		for (size_t j = 0; j < (size_t)freq_count; j++)
			if (freqs[j].first == set[i])
				found = j;
		if (found == -1)
			freqs[freq_count++] = (IntPair) {
				.first = set[i],
				.second = 1,
			};
		else
			freqs[found].second++;
	}
	return (IntMap) {
		.pairs = freqs,
		.count = freq_count,
	};
}

static bool has_same(dice_t dice, int count) {
	IntMap counts = frequencies(dice.faces);
	for (size_t i = 0; i < (size_t)counts.count; i++)
		if (counts.pairs[i].second == count)
			return true;
	return false;
}

static bool has_same_or_bigger(dice_t dice, int count) {
	IntMap counts = frequencies(dice.faces);
	for (size_t i = 0; i < (size_t)counts.count; i++)
		if (counts.pairs[i].second >= count)
			return true;
	return false;
}

static bool is_category(dice_t dice, category_t category) {
	if (category == FULL_HOUSE)
		return has_same(dice, 2) && has_same(dice, 3);
	else if (category == FOUR_OF_A_KIND)
		return has_same_or_bigger(dice, 4);
	else if (category == LITTLE_STRAIGHT)
		return is_little_straight(dice);
	else if (category == BIG_STRAIGHT)
		return is_big_straight(dice);
	else if (category == YACHT)
		return is_yacht(dice);
	else
		return true;
}

static int find_count_same_or_bigger(dice_t dice, int count) {
	IntMap freqs = frequencies(dice.faces);
	for (size_t i = 0; i < (size_t)freqs.count; i++)
		if (freqs.pairs[i].second >= count)
			return freqs.pairs[i].first;
	assert(false);
}

int score(dice_t dice, category_t category) {
	if (!is_category(dice, category))
		return 0;

	switch (category) {
	case ONES:
	case TWOS:
	case THREES:
	case FOURS:
	case FIVES:
	case SIXES: return category * count(dice, category);
	case CHOICE:
	case FULL_HOUSE: return sum(dice);
	case FOUR_OF_A_KIND: return find_count_same_or_bigger(dice, 4) * 4;
	case LITTLE_STRAIGHT:
	case BIG_STRAIGHT: return 30;
	case YACHT: return 50;
	}
	assert(false);
}
